Proto extensions spec (#7584)
microsoft/terminal: doc/specs/Proto extensions-spec.md

Currently, Ubuntu/WSL/Powershell etc are unable to provide any specifications for how they want
their profiles in Terminal to look - only we and the users can modify the profiles. We want to provide
these installations with the functionality to have a say in this, allowing them to specify things like
their icon, their font and so on. However, we want to maintain that the user has final say over all of
these settings. 

## Solution Design

Currently, when we load the settings we perform the following steps (this is a simplified description,
for the full version see the spec for cascading default + user settings):

1. Generate profiles from the defaults file
2. Generate profiles from the dynamic profile generators
3. Layer the user settings on top of all the profiles created in steps 1 and 2
4. Validate the settings

To allow for installations to add in their snippets of json, I propose the addition of a new step
in between 2 and 3:

1. Generate profiles from the defaults file
2. Generate profiles from the dynamic profile generators
3. **Incorporate the additional provided json stubs** - these stubs could be:
   1. modifications to existing profiles

>    2. full profiles

Reviewer: ```suggestion
   2. additions of full profiles
```
nit

>    3. colour schemes/modifications to colour schemes

Reviewer: ```suggestion
   3. additions of colour schemes
   4. modifications to colour schemes
```

4. Layer the user settings on top of all the profiles created in steps 1 through 3
5. Validate the settings

### Specifications of the json stubs

> As written above, the json stubs could be modifications to existing profiles, full profiles, or colour schemes.

Reviewer: ```suggestion
As written above, the json stubs could be modifications to or additions of profiles or colour schemes.
```
Even then, this sentence still reads weird. The "modifications to" can be attached to "full profiles" which makes it a bit confusing. You also don't mention what you're doing to colour schemes (adding or modifying? Or both!)

This is more of a grammar nit though.

> #### Modifications to existing profiles

Reviewer: For each of these sub-sections, please include an example of the JSON stub that the creator would submit. It'll help make this more clear.

For modifications to existing profiles, the json stub would need to indicate which profile it wishes to modify.
A straightforward way to do this would be by providing the name of the profile. Since this modification would
occur at the step _before_ any user changes to names are layered, the names of the profiles are consistent and
can be used for identification. For example, the name of a profile generated by the WSL profile generator is
simply the argument to "wsl.exe -d \<name\>", the name of a profile generated by the Powershell profile generator

> would be "Powershell Core" or "Powershell (ARM)" etc depending on the version.

Reviewer: I don't like the idea of indexing by the name. Because what if a stub is intended to rename a profile. Consider a case where the user has "Ubuntu 20.04" installed. Shouldn't a stub be able to rename it just plain "Ubuntu"?

We have GUIDs for unique identifiers. Why not just use those?

Reviewer: I think we consider 2 profiles to be the same if the guids and source match.
Source: Profile.cpp - `bool Profile::ShouldBeLayered(const Json::Value& json) const`

We might run into the case where multiple json stubs modify the same profile and so they override each other. We
will need some sort of hierarchy to determine which stub to ultimately use. My initial thought is that we could
check the owner of the json file and see if it matches the profile they are trying to modify (so something like
only Powershell can modify the Powershell profiles and only Ubuntu can modify the Ubuntu profile etc). Independent

> parties will need to create full profiles with the desired commandline argument instead.

Reviewer: If two stubs try modifying the same profile, I see two options here:
1. The last stub that's imported wins (that's what we generally do now)
2. Layer the stubs onto the existing profile in that order

I think option 1 makes the most sense here. We already do that, in general. Layering can result in bad combos (i.e.: foreground=white, background=white)

Author: For now, we are just going to go with applying _all_ the changes

> #### Full profile stubs

Reviewer: After a user installs a profile from a stub, the user's profiles.defaults get applied. Is there a way for the user to apply more customizations onto the newly created profile?

Author: Yes

> As in the case of the dynamic profile generator, if we create a profile that did not exist before (i.e. does not
exist in the user settings), we need to add the profile to the user settings file and re-save that file. 

So the user can just add more customizations as normal

Technically, full profile stubs do not need to contain anything (they could just be '\{\}'). However we should
have some qualifying minimum criteria before we accept a stub as a full profile. I suggest that we only create
new profile objects from stubs that contain at least the following

* A name (and this name _should not_ match the names of existing profiles as described above, because that would cause us to interpret this profile as modifying an existing one)
* A commandline argument
* A unique GUID - this is so we avoid conflicts if several different creators name their profile the same thing

> As in the case of the dynamic profile generator, if we create a profile that did not exist before (i.e. does not

Reviewer: What does this stamped stub look like in settings.json? Will it have a "source" to denote that it came from this stub? Is it just the stub copied/pasted directly? Or will it look small/lightweight like the dynamically generated profiles?

Author: It will be copied/pasted directly (just as if the user typed it up themselves), which would allow the user to customize it further as well. We might add a source to it too - pending discussion on that.

Author: It will be lightweight like the dynamically generated profiles and it will have a source!

exist in the user settings), we need to add the profile to the user settings file and re-save that file. 

#### Colour schemes

> Once again, we could run into the problem of several json stubs wanting to modify the same colour scheme. We

Reviewer: ColorSchemes are special in that their only identifier is "name". As with profiles, I see 2 options here:
1. The last stub that's imported wins (that's what we generally do now)
2. Layer the stubs onto the existing color scheme in that order
Again, my vote is for option 1 since layering could result in bad combos (i.e. white foreground on white background).

Reviewer: Since `name` is the unique identifier, is there any way to rename a color scheme?

Author: We are just going to allow same name colour schemes to override each other for now

will require a hierarchy here as well. Though since the 'owner' of a colour scheme is a lot more dubious, I wonder
if we should only allow creation of new colour schemes for now and ignore requests to modify existing colour

> schemes. 

Author: And if someone wants to add a new colour scheme, we should probably append their unique identifier to the name (so we don't allow someone to lay claim to the name "Best Colour Scheme Ever" or something)

Reviewer: I don't like the idea of appending a unique identifier to the name, unless we can rename it. If that's the case, please go into more details here. Maybe we need to add a "source" to the imported color scheme?

Author: That would mean we would need to revamp how we interpret colour schemes - for example if someone adds a colour scheme named "Campbell", when we parse out what that "Campbell" means we will also need to match the source before we can apply the scheme (and that means we will also need to define a behaviour for when the source is unspecified). 

I feel like appending a unique identifier to the name gives us the same functionality without needing to mess with everywhere else a colour scheme is interpreted.

Author: For now, we will just let multiple colour schemes with the same name override each other.

> ### Location(s) of the json files

Reviewer: Could stubs be saved to an online location (like a url)?

Author: Perhaps potentially? That would probably involve us making an internet connection to obtain the data and then using that. However I'd say that is beyond the scope of this spec.

For apps that are installed 'traditionally', their installer could simply add a json file to a specific local folder
(specified by us) that we will look through every time we generate profiles. This is also where users can add
their own json files for Terminal to generate profiles from. 

For apps that are installed through something like the Microsoft Store, they should add their json stub(s) to

> an app extension. We will then obtain them during profile generation. 

Reviewer: What does this look like? How do we detect the app extension (or is it their job to update us?)

> ## UI/UX Design

Reviewer: Either here, or at the beginning of "Solution Design", please include the process of...
- How a maintainer creates a stub...
   - What's the structure of the stub (reference "Specifications of the json stubs")
   - What do they do with the stub once it's created? Do they upload it somewhere?
   - Can the creator modify the existing stub?
- How a user adds a one of these stubs...
   - Do they download the stub to a folder of their choice and link it in their settings.json?
   - Does the stub have to be downloaded to a specific directory?
   - What modifications does the user have to make to settings.json (if any) to apply these stubs?

Author: _Should_ be clearer now! Let me know if more detail is required

This feature will allow other installations a level of control over how their profiles look in Terminal. For example,
if Ubuntu gets a new icon or a new font they can have those changes be reflected in Terminal users' Ubuntu profiles.

Furthermore, this allows users an easy way to share profiles they have created - instead of needing to modify their
settings file directly they could simply download a json file into a specific folder. 

## Capabilities

### Accessibility

This change should not affect accessibility.

### Security

Opening a profile causes its commandline argument to be automatically run. Thus, if malicious modifications are made
to existing profiles or new profiles with malicious commandline arguments are added, users could be tricked into running

> things they do not want to.

Author: This could already happen even now though (someone could replace the settings file), so I don't think this adds _that_ much risk

### Reliability

This should not affect reliability - most of what its doing is simply layering json which we already do. 

### Compatibility

Again, there should not be any issues here - the user settings can still be layered after this layering for the user
to have the final say. 

### Performance, Power, and Efficiency

Looking through the additional json files could negatively impact startup time. 

## Potential Issues

Cases which would likely be frustrating:

* An installer dumps a _lot_ of json files into the folder which we need to look through

> ## Future considerations

Reviewer: How does this interact with a future extensions model? Or is this going to be the only way to add/modify Profiles and ColorSchemes?

Author: This should not affect a future extensions model - and we will need this eventually regardless!

How would this affect the Settings UI? 

> 

Author: @carlos-zamora, would like to hear your input here!

Reviewer: I'm still a bit confused as to what changes we're making to the settings.json. So I'll hold off on answering this for now.

You may want to look at this for now: https://github.com/microsoft/terminal/blob/cinnamon/spec-settings-ui/doc/specs/%231564%20-%20Settings%20UI/design.md

We are planning on adding a "Marketplace" nav item. So it makes sense that this would go there. But we can probably hold off on a full design until after this spec gets merged in.

Author: We aren't making changes to the structure of settings.json - only the potential additions of more profiles similar to the way dynamic profiles are added. So I want to say that it probably wouldn't affect the settings UI but I will leave it to you to confirm!

Reviewer: Looks like the only impact this would have is how we display "source" in the Settings UI. So I'll just keep this in mind as we go. No updates needed on your end :)

    "commandline": "powershell.exe",
    "antialiasingMode": "aliased",
    "fontWeight": "bold",
    "scrollbarState": "hidden"
}
```

#### Colour schemes

Once again, we could run into the problem of several json stubs wanting to modify the same colour scheme. We
will require a hierarchy here as well. Though since the 'owner' of a colour scheme is a lot more dubious, I wonder
if we should only allow creation of new colour schemes for now and ignore requests to modify existing colour
schemes. 

### Creation and location(s) of the json files

#### For apps installed through Microsoft store (or similar)

For apps that are installed through something like the Microsoft Store, they should add their json file(s) to
an app extension. During our profile generation, we will probe the OS for app extensions of this type that it
knows about and obtain the json files.

#### For apps installed 'traditionally' and third parties/independent users

For apps that are installed 'traditionally', their installer could simply add a json file to a specific local folder
(specified by us) that we will look through every time we generate profiles. This is also where independent users will

> add their own json files for Terminal to generate/modify profiles. 

Reviewer 2: Something is nagging at me that we should use the same COM/WinRT registration plan we're going to use for Default App and probably Real Extensions to do this. But maybe I'm just drinking the COM kool-aid. @DHowett, thoughts?

Reviewer 2: I was drinking the kool-aid. We discussed this offline. I rescind my comment.

#### Creation of colour schemes

As with full profiles, we will have some qualifying criteria for what we accept as full colour schemes: color schemes
must have
* A name
* A background
* A foreground

This will cause the issue of colour schemes being overridden if there are many creations of a colour scheme with the
same name. Since for now all we have to uniquely identify colour schemes *is* the name, we will just let this be.

Here is an example of a json stub that contains a colour scheme:

```js
{
    "name": "Postmodern Tango Light",
    "background": '#61D6D6',
    "foreground": '#E74856'
}
```

This stub will *not* show up in the users settings file, similar to the way our default colour schemes do not show up.

### Creation and location(s) of the json files

#### For apps installed through Microsoft store (or similar)

For apps that are installed through something like the Microsoft Store, they should add their json file(s) to
an app extension. During our profile generation, we will probe the OS for app extensions of this type that it
knows about and obtain the json files or objects. These apps should should use msappx for us to know about

> their extensions.

Author: @miniksa would like your input here on the specifics of probing the OS for app extensions!

5. Validate the settings

### Specifications of the json stubs

As written above, the json stubs could be modifications to existing profiles, additions to full profiles, or additions of colour schemes.

To allow modifications/additions of several profiles in one file and to allow the addition of several colour schemes in one json file,
these stubs should be put into lists in the json file: one list named ```profiles``` and the other list named ```schemes```. For example:

```js
{
    "profiles": [ modifications and additions of profiles go here ],
    "schemes": [ additions of colour schemes go here ]
}
```

An example of a full json file with these fields filled out is shown at the end of this section. 

#### Modifications to existing profiles

The main thing to note for modification of existing profiles is that this will only be used for modifying the
default profiles (cmd/powershell) or the dynamically generated profiles. 

For modifications to existing profiles, the json stub would need to indicate which profile it wishes to modify. It will

> do this by providing the corresponding guid in the `"updates"` field of the json stub.

Reviewer 3: this might need to be called out in the spec better.

There's a distinct difference between the fragment...

```json
{
    "guid": "{65f3c7d5-f241-483c-b7c4-982b5361c318}",
    "name": "Ubuntu 28.02 Xylophonetic Xorn",
}
```

...and...

```json
{
    "updates": "{65f3c7d5-f241-483c-b7c4-982b5361c318}",
    "name": "Ubuntu 28.02 Xylophonetic Xorn",
}
```

The first one will create a new profile, even if the WSL distro generator didn't attest it.

We could use `source`, but that's baking a contract into another static file and holds Ubuntu back if we or a community member eventually adds "Microsoft.Terminal.WslButBetterBecauseItCanFindIconsButTheUserMayWantToDisableItAndFallBackToTheOldOneBecauseItIsExpensive" (sorry, mouthful).

We don't have a good story for sources overriding eachother, and we don't have a good system for overriding _regardless of source_.

This is an implementation of, effectively, "If-Match" from HTTP. "If you have one, I am interested. If you do not, leave me alone."

> do this by providing the corresponding guid in the `"updates"` field of the json stub.

Reviewer 4: AHHH okay. I see how that works. So for apps where we've already got a dynamic profile generator (WSL distros), they're going to want to use `updates`, but pretty much everyone else will want to use `guid`. Got it.

Note that currently, we generate a GUID for dynamic profiles using the "initial" name of the profile (i.e. before
any user changes are applied). For example, the "initial" name of a WSL profile is the \<name\> argument to
"wsl.exe -d \<name\>", and the "initial" name of a Powershell profile is something like "Powershell (ARM)" - depending
on the version. Thus, the stub creator could simply use the same uuidv5 GUID generator we do on that name to obtain the
GUID. Then, in the stub they provide the GUID can be used to identify which profile to modify.

Naturally, we will have to provide documentation on how they could generate the desired GUID themselves. In that documentation,
we will also provide examples showing how the current GUIDs are generated for clarity's sake. 

We might run into the case where multiple json stubs modify the same profile and so they override each other. For the initial implementation, we

> are simply going to apply _all_ the changes. Eventually, we will probably want some sort of hierarchy to determine

Reviewer 3: (Sorry, can't see if this is addressed outside the snippet above.)
Yeah -- we talked about "altitudes" and if we have an altitude system we could say "all dynamic generators run at 50, user profiles at 100, and fragments can define where they run"

> are simply going to apply _all_ the changes. Eventually, we will probably want some sort of hierarchy to determine

Author: Updated this part based on our discussion - we are decidedly going with no defined ordering

For apps that are installed through something like the Microsoft Store, they will need to declare themselves to
be an app extension or write a separate app extension. In the app extension, they will need to create a public
folder and insert their json files into it. The specifics of how they can do this are provided in the
[Microsoft Docs](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/how-to-create-an-extension) for how
to create an app extension, and I will replicate the necessary section here. 

In the appxmanifest file of the package:

```
<Package
  ...
  xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3"
  IgnorableNamespaces="uap uap3 mp">
  ...
    <Applications>
      <Application Id="App" ... >
        ...
        <Extensions>
          ...
          <uap3:Extension Category="windows.appExtension">
            <uap3:AppExtension Name="Microsoft.com.Terminal"
                               Id="<id>"
                               PublicFolder="Public">
            </uap3:AppExtension>
          </uap3:Extension>
        </Extensions>
      </Application>
    </Applications>
    ...
</Package>
```

> Note that the name field **must** be `Microsoft.com.Terminal` for us to detect this extension. The `Id` field

Author: The [code sample](https://github.com/MicrosoftDocs/windows-topic-specific-samples/blob/MathExtensionSample/MathExtension/Package.appxmanifest) from the [app extension docs](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/how-to-create-an-extension) has `Microsoft.com.MathExt`, I just followed that style - do we know which one is the standard?

> Note that the name field **must** be `Microsoft.com.Terminal` for us to detect this extension. The `Id` field

Reviewer 4: Oh my god the code sample doesn't line up with the docs
![image](https://user-images.githubusercontent.com/18356694/95103698-8cdfcd00-06fa-11eb-9e47-e34482ff62f3.png)
![image](https://user-images.githubusercontent.com/18356694/95103728-96693500-06fa-11eb-8be5-c83e24204417.png)

vs

https://github.com/MicrosoftDocs/windows-topic-specific-samples/blob/a68d04dc12950acdedf1a804d0f90fd07424f370/MathExtension/Package.appxmanifest#L27-L38

> Note that the name field **must** be `Microsoft.com.Terminal` for us to detect this extension. The `Id` field

Author: Hm I'll assume the docs are the standard and change it then

can be filled out as the app wishes. The `PublicFolder` field should have the name of the folder, relative to
the package root, where the `json` files they wish to share with us are stored (this folder is typically named
`Public` but can be renamed as long as it matches the relevant folder).

> During our profile generation, we will probe the OS for app extensions with the name `Microsoft.com.Terminal`

Author: We could also say that json files for settings should be put into a `Fragments` subdirectory within the `public` folder, so in the future if apps want to share more data with us they can just put it elsewhere in the `public` folder rather than declare a new app extension (correspondingly, we won't need to declare ourselves a new app extension host either). Which do you prefer?

> During our profile generation, we will probe the OS for app extensions with the name `Microsoft.com.Terminal`

Reviewer 4: I'm thinking a `fragments` subdirectory is the best plan. Synergizes well with my request below too.

and obtain the json files stored in the public folders of those app extensions.

#### For apps installed 'traditionally' 

> For apps that are installed 'traditionally', there are 2 cases. The first is that this installation is for all

Author: The extension app, I will make this clearer

the users of the system - in this case, the installer should add their json files to the global folder:

```

> C:\ProgramData\Microsoft\Windows Terminal\Fragments

Author: > We could also say that json files for settings should be put into a Fragments subdirectory within the public folder, so in the future if apps want to share more data with us they can just put it elsewhere in the public folder rather than declare a new app extension (correspondingly, we won't need to declare ourselves a new app extension host either). Which do you prefer?

```

Note: `C:\ProgramData` is a [known folder](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb776911(v=vs.85))
that apps should be able to access. It will be on us to create the folder `Windows Terminal` in `C:\ProgramData\Microsoft\Windows`
for this to happen. 

In the second case, the installation is only for the current user. For this case, the installer should add the
json files to the local folder:

```
C:\Users\<user>\AppData\Local\Microsoft\Windows Terminal\Fragments
```

> We will look through both folders mentioned above during profile generation. 

Author: Right now the order is `local` -> `global` -> `app`. Though if we eventually go for something like an 'altitudes' system then it wouldn't matter which directory is scanned first since it will be about the json stub itself.

#### The source field

> Currently, we allow users an easy way to disable/enable profiles that are generated from certain sources. For

Author: Currently, yes

> Currently, we allow users an easy way to disable/enable profiles that are generated from certain sources. For

Reviewer 4: Is that something we want to consider for this scenario? Lets use #7774 as an example. Let's imagine that VS has left behind fragments for your VS installs, but I want to tidy that up and hide all the VS ones. Now I can't get any other proto-extension fragments from extension apps of the same class (local/global/app).

> Currently, we allow users an easy way to disable/enable profiles that are generated from certain sources. For

Author: Updated this based on our discussion!

## UI/UX Design

This feature will allow other installations a level of control over how their profiles look in Terminal. For example,
if Ubuntu gets a new icon or a new font they can have those changes be reflected in Terminal users' Ubuntu profiles.

## Capabilities

### Accessibility

This change should not affect accessibility.

### Security

Opening a profile causes its commandline argument to be automatically run. Thus, if malicious modifications are made
to existing profiles or new profiles with malicious commandline arguments are added, users could be tricked into running
things they do not want to.

### Reliability

This should not affect reliability - most of what its doing is simply layering json which we already do. 

### Compatibility

Again, there should not be any issues here - the user settings can still be layered after this layering for the user
to have the final say. 

### Performance, Power, and Efficiency

Looking through the additional json files could negatively impact startup time. 

## Potential Issues

* An installer dumps a _lot_ of json files into the folder which we need to look through.

> * When a `.json` files is deleted, any new profiles that were generated from it remain in the user's settings file (though they no longer appear in the tab dropdown).

Author: That already happens - say I get a new profile from a json file, then delete the file, that new profile will not appear in the tab dropdown anymore (the stub sticks around in the user's settings file though, but that is also how the dynamic profiles work)

> * When a `.json` files is deleted, any new profiles that were generated from it remain in the user's settings file (though they no longer appear in the tab dropdown).

Reviewer 4: Oh but I believe with dynamic profiles, we see that the user had a stub for a dynamic profile, but also that we _didn't_ generate the profile which that stub corresponds to. I _believe_ in that case, we actually insert `"hidden": true` into the user's settings for that profile, to auto-hide it. We should probably do that for the stubs we're generating here as well.

> * When a `.json` files is deleted, any new profiles that were generated from it remain in the user's settings file (though they no longer appear in the tab dropdown).

Author: Chatted with @DHowett , apparently we _don't_ change the `hidden` field in the user's settings in that case

> * When a `.json` files is deleted, any new profiles that were generated from it remain in the user's settings file (though they no longer appear in the tab dropdown).

Reviewer 3: Yeah, we've only ever patched _one_ individual settings key, and that's `$schema`. For all the rest, we only insert profile blobs and then _ignore them_ when they don't exist.

If we set hidden:true, it would be hidden even after the profile came back. Consider how the WSL generator apparently fails to find WSL every other morning for some folks!

> * When a `.json` files is deleted, any new profiles that were generated from it remain in the user's settings file (though they no longer appear in the tab dropdown).

Reviewer 4: Oh, my b. Must've been remembering an earlier version of that spec then. As long as the profile's gone, I'm cool with this 👍

files. Once we implement this feature, we will need to provide documentation on this for app developers. 

#### For apps installed through Microsoft store (or similar)

For apps that are installed through something like the Microsoft Store, they will need to declare themselves to
be an app extension or write a separate app extension. In the app extension, they will need to create a public
folder, and create a subdirectory within that folder called `Fragments`. The json files should be inserted into the
`Fragments` subdirectory. The specifics of how they can declare themselves as an app extension are provided in the
[Microsoft Docs](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/how-to-create-an-extension), and I will
replicate the necessary section here. 

In the appxmanifest file of the package:

```
<Package
  ...
  xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3"
  IgnorableNamespaces="uap uap3 mp">
  ...
    <Applications>
      <Application Id="App" ... >
        ...
        <Extensions>
          ...
          <uap3:Extension Category="windows.appExtension">

>             <uap3:AppExtension Name="com.Microsoft.Terminal"

Reviewer 2: Please use lowercase for all refs `com.microsoft.terminal`.

                               Id="<id>"
                               PublicFolder="Public">
            </uap3:AppExtension>
          </uap3:Extension>
        </Extensions>
      </Application>
    </Applications>
    ...
</Package>
```

> Note that the name field **must** be `com.Microsoft.Terminal` for us to detect this extension. The `Id` field

Reviewer 2: Please also add a subcategory like `com.microsoft.terminal.settingsext` because I intend to use `com.microsoft.terminal.host` for the 3rd party terminal hosts and `com.microsoft.console.host` for 3rd party console host registrations.

Reviewer 2: Urgh, I used "com.microsoft.windows.terminal.host" in mine, so maybe "com.microsoft.windows.terminal.settingsext" or something. We should just be consistent with each other.

Reviewer 4: > setting sext

Maybe just `settings`?

For modifications to existing profiles, the json stub would need to indicate which profile it wishes to modify.
Note that currently, we generate a GUID for dynamic profiles using the "initial" name of the profile (i.e. before
any user changes are applied). For example, the "initial" name of a WSL profile is the \<name\> argument to
"wsl.exe -d \<name\>", and the "initial" name of a Powershell profile is something like "Powershell (ARM)" - depending
on the version. Thus, the stub creator could simply use the same uuidv5 GUID generator we do on that name to obtain the
GUID. Then, in the stub they provide the GUID can be used to identify which profile to modify.

Naturally, we will have to provide documentation on how they could generate the desired GUID themselves. In that documentation,
we will also provide examples showing how the current GUIDs are generated for clarity's sake. 

We might run into the case where multiple json stubs modify the same profile and so they override each other. For the initial implementation, we
are simply going to apply _all_ the changes. Eventually, we will probably want some sort of hierarchy to determine

> an order to which changes are applied.

Author: We will layer all the stubs onto the profile in whatever order we find the stubs in (so there might be some overriding but we are not going to deal with that for now)

Here is an example of a json stub that modifies an existing profile (specifically the Azure cloud shell profile):

```js
{
    "guid": "{b453ae62-4e3d-5e58-b989-0a998ec441b8}",
    "fontSize": 16,
    "fontWeight": "thin"
}
```

**NOTE**: This will *not* change the way the profile looks in the user's settings file. The Azure cloud shell profile
in their settings file (assuming they have made no changes) will still be as below. 

```js
{
    "guid": "{b453ae62-4e3d-5e58-b989-0a998ec441b8}",
    "hidden": false,
    "name": "Azure Cloud Shell",
    "source": "Windows.Terminal.Azure"
}
```

> However, the user is free to make changes to it as usual and those changes will have the 'final say'. 

Reviewer: In this example, if the user has "fontSize" defined in profiles.defaults, which one is applied? (might be worth adding it to the numbered list earlier too).

Author: The user's one! Its still considered a 'user setting' so it will get applied last

The source field value is dependent on how we obtained this json file, and will be touched on in more detail later in the
spec (see "Creation and location(s) of the json files" - "The source field"). 

Here is an example of a json stub that contains a full profile:

```js
{
    "guid": "{a821ae62-9d4a-3e34-b989-0a998ec283e6}",
    "name": "Cool Profile",
    "commandline": "powershell.exe",
    "antialiasingMode": "aliased",
    "fontWeight": "bold",
    "scrollbarState": "hidden"
}
```

When this profile gets added back to the user's settings file, it will look similar to the way we currently add
new dynamic profiles to the user's settings file. Going along with the example above, the corresponding
json stub in the user's settings file will be:

```js
{
    "guid": "{a821ae62-9d4a-3e34-b989-0a998ec283e6}",
    "name": "Cool Profile",
    "hidden": "false",
    "source": "local"
}
```
Again, the user will be able to make changes to it as they do for all other profiles. 

#### Creation of colour schemes

As with full profiles, we will have some qualifying criteria for what we accept as full colour schemes: color schemes
must have
* A name
* A background

> * A foreground

Reviewer: Personally, don't like the idea of partially defined color schemes. I'd say every field is required (except selectionBackground?).

Especially because foreground/background can just be defined on a profile (stub).

Author: I'm not sure about needing to specify _every_ field though - if someone wants to create a fairly simple colour scheme I don't think they should need to specify BrightCyan, BrightGreen etc. 

How about, at the minimum, they should specify: foreground, background, cursor colour and selection background?

This will cause the issue of colour schemes being overridden if there are many creations of a colour scheme with the
same name. Since for now all we have to uniquely identify colour schemes *is* the name, we will just let this be.

Here is an example of a json stub that contains a colour scheme:

```js
{
    "name": "Postmodern Tango Light",
    "background": '#61D6D6',
    "foreground": '#E74856'
}
```

> This stub will *not* show up in the users settings file, similar to the way our default colour schemes do not show up.

Reviewer: (thinking out loud, not necessarily for the spec) This'll be interesting for the Settings UI, because it's valid to _reference_ this color scheme, but not _modify_ it. The Color Schemes page may have to either...
- remove ColorSchemes from stubs
- present them, but put a note that it's imported and not modifiable
- present them, but when the user modifies one, the modified version gets copied to the settings.json

Author: None of the current colour schemes we have show up in the settings file though, how does the settings UI deal with those?

Reviewer: Ah. So `GlobalAppSettings` currently has a reference to _all_ of the color schemes. This can't be changed because that's how we look up what a color scheme name actually maps to. The Settings UI reads/edits that instead of the actual JSON directly.

So instead, we need a way to mark where each color scheme came from (like `source` in `Profile`). That way, when the Settings UI presents the Color Schemes page, we handle stub-generated color schemes differently (the list of ideas above).

We _could_ also introduce another top-level navigation item that lets users browse/copy/modify anything generated from stubs. But I'm less of a fan of that idea because it breaks up where all of the color schemes can be found.
